app/components/modelling/baseModel.jsx

```diff
@@ -0,0 +1,147 @@
+    // set the texture of the image we are modelling
+    this.setBaseTexture = this.setBaseTexture.bind(this);
+    this.parse = this.parse.bind(this); // requires overwrite
+    this.getScore = this.getScore.bind(this); // (optional)
+
+    // save info on the zooniverse subject
+    this.baseTexture = null;
+    this.imageMetadata = Object.assign({}, metadata);
+    this.size = [canvas.height, canvas.width];
+    this.differenceData = new Uint8Array(this.size[0] * this.size[1] * 4);
+
+    // instantiate the regl instance with the provided canvas
+    this.regl = reglBase({
+      canvas,
+      attributes: { preserveDrawingBuffer: true }
+    });
+    // create a texture which is used for post-processing and difference calculation
+    this.pixels = this.regl.texture();
+    const postProcessingArray = this.getPostProcessingFunc(this.regl);
+
+    this.postProcessingFunc = () => {
+      if (this.baseTexture !== null) {
+        if (typeof this.imageMetadata.psf === 'undefined') {
+          /* eslint-disable prefer-spread */
+          const psf = Array.apply(null, Array(121)).map(Number.prototype.valueOf, 0);
+          /* eslint-enable prefer-spread */
+          psf[60] = 1;
+          this.imageMetadata.psf = psf;
+        }
+        postProcessingArray.map((i) => {
+          this.pixels({
+            copy: true
+          });
+          i({
+            texture: this.pixels,
+            metadata: this.imageMetadata,
+            imageTexture: this.baseTexture
+          });
+        });
+      }
+    };
+
+    // call the getModel function, this will be extended for each model and returns
+    // a list of component objects TODO: model schema?
+    this.model = this.getModel(this.regl);
+    this.toRender = [];
+  }
+  update(annotation) {
+    // this function parses the new annotation and then triggers a render.
+    // update the render funtions
+    this.toRender = this.parse(annotation);
+    this.regl.poll();
+    this.regl.clear({
+      color: [0, 0, 0, 1]
+    });
+    this.pixels({
+      copy: true
+    });
+    for (let i = 0; i < this.toRender.length; i += 1) {
+      this.toRender[i][0](Object.assign({ texture: this.pixels }, this.toRender[i][1]));
+      this.pixels({
+        copy: true
+      });
+    }
+    this.pixels({
+      copy: true
+    });
+    this.postProcessingFunc();
+  }
+  kill() {
+    this.regl.destroy();
+  }
+  setBaseTexture(image) {
+    try {
+      this.baseTexture = this.regl.texture(image);
+    } catch (err) {
+      if (err.name === 'SecurityError') {
+        // TODO: shouldn't really be using alert here. Does Zooniverse have a modal?
+        alert(
+          (resolve, reject) =>
+            (
+              <div className="content-container">
+                <Markdown className="classification-task-help">
+                  CORS error encountered, please make sure you are using Google Chrome and try refreshing the page
```

Comment:
Is this CORS error still happening, I thought you fixed it?  If it is fixed this entire `alert` can go away.

Reply:
CORS will still trigger on browsers that don't support `crossorigin="anonymous"` on image tags (i.e. most recent safari, I think). The bug I spotted and fixed was displaying as CORS but actually something else that I hadn't been able to catch.